manpages/imod2obj.man
.na
.nh
.TH imod2obj 1 2.00 BL3DEMC
.SH NAME
imod2obj \- Make a OBJ model from an IMOD model file.
.SH SYNOPSIS
imod2obj  [options]  IMOD_model  obj_file  [mtl_file]
.SH DESCRIPTION
Convert an IMOD model file to the Wavefront .obj text file format which
a very simple but common format for interchange of 3D meshes between
various 3D graphics programs.  OBJ files are generally much more limited
than VRML 2.0 and other 3D file formats as there is no object hierarchy and they
lack support for primitives like spheres. As a result, all spheres are
converted  to meshes and all meshes are listed in a flat array of "groups".
OBJ files do not store material properties, but they do allow materials to
be referenced in a separate "Material Template Library" (MTL) file which is
generated if you include a second output file (mtl_file.mtl) on the end of
the command. While OBJ files are widely used, not all computer programs will
know how to interpret MTL file and thus it's likely your color and material
information will be lost. 
.P
The model produced will generally have similar attributes to those set in
3dmod's ModelView before the conversion. Specifically, objects that are turned
off will not be included in the OBJ file. OBJ files can and do not allow
contours to be rendered as contours, thus only objects which have been meshed
will appear in the OBJ file.
.SH OPTIONS
.TP 
.B -l
Use low-resolution meshes, if any are available.  The default is to use the
highest resolution meshes contained in the model.  Low resolution meshes can
be produced with the 
.B -l
option to imodmesh(1).
.TP 
.B -a
Output all objects (by default those switched off are omitted).
.TP 
.B -r
Rotates the entire model such that the Z axis becomes up in Y.
This feature is useful when exporting to animation programs
.TP 
.B -n
Output normals (off by default as most 3d programs which import OBJ can
generate their own normals if missing).
.TP
.B -s \fI#\fR
Specifies number of segments to render per sphere - the more segments the
nicer the sphere, but the more faces and thus longer to load. Think of this
number as the number of lines of longitude (the ones that go up and down).
The default value is 8 and the number of (horizontal) lines of latitude is
made half this value.
.TP
.B -i
Print spheres as icosahedrons (a surface with 20 equilateral triangles)
instead of "standard spheres".
.SH AUTHORS
Andrew Noske
.SH SEE ALSO
3dmodv(1), imodmesh(1), imod2vrml(1), imod2vrml2(1)
.SH BUGS
Email bug reports to mast@colorado.edu, and CC to andrew_noske<AT>yahoo<DOT>com.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)


imodutil/imod2obj.cpp
/*
 *  imod2obj.cpp - convert IMOD model to (wavefront) OBJ file
 *  
 *  Author:     Andrew Noske
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 *  Info on OBJ: http://www.andrewnoske.com/wiki/index.php?title=OBJ_file_format
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id: imod2obj.cpp
 *  Log at end
 */

//############################################################

#include <stdio.h>
#include <string.h>
#include <math.h>
#include "imodel.h"

//## GLOBAL VARIABLES:

static int  numObjs     = 0;		// tallies the number of objects turned on
static int  numVertices = 0;		// tallies the number of vertex points added
static int  numFaces    = 0;		// tallies the number of faces added
static int  numNormals  = 0;		// tallies the number of normals added
static int  numSpheres  = 0;		// tallies the number of spheres printed

static int lowRes         = 0;		// if we want to use low res version
static int printAllObjs   = 0;		// print all objects (not just those on)
static int rotateModel    = 0;		// "rotate" (flip Y and Z axis)
static int printMatFile   = 0;		// generate a .mtl file
static int printNormals   = 0;		// print normals
static int flipNormals    = 0;    // flip normals
static int sphereSegments = 8;		// number of segments per sphere
static int useIcosahedrons= 0;		// draw isohedrons instead of "standard" sphere meshes
static int onlyScatSpheres= 0;	  // only print spheres for scattered objects

// TODO: Add labels...


//## FUNCTION DECLARATION:

void imod_to_obj(Imod *imod, FILE *fout, char *outFileName);
static void printObject(Imod *imod, int ob, FILE *fout);
static void printObjSafeUniqueObjectName(Imod *imod, int ob, FILE *fout );
static void printMesh(Imod *imod, int ob, FILE *fout);
static void printScatContours(Imod *imod, int ob, FILE *fout);
static void printSphere(Ipoint pt, float radius, int segments, FILE *fout);
static void printSphereAsIcosahedron(Ipoint pt, float radius, FILE *fout);

void imod_to_mtl(Imod *imod, FILE *fout);
static void printMaterial(Imod *imod, int ob, FILE *fout);

//############################################################

//------------------------
//-- Usage function - used to write out a small set of instructions in the console
//-- window when the user types "imod2obj -h" or gets an argument wrong.

static void usage(int error)
{
  printf("\nConverts an imod model to OBJ (wavefront object) file format and\n");
	printf("   (if specified) a matching MTL (material template library) file.\n\n");
  printf("Usage: imod2obj [options] <input.mod> <output.obj> [output.mtl]\n");
	printf("Options:\n");
  printf("       -l     output low-resolution meshes (if any exist).\n");
	printf("       -a     output all objects (by default any switched off are omitted)\n");
	printf("       -r     rotates the model by 90 pitch by effectively flipping the \n");
	printf("              Y and Z axis - a better orientiation making movies\n");
	printf("       -n     output normals (off by default as most 3d programs which \n");
	printf("              import OBJ can generate their own normals if missing).\n");
	printf("       -f     flips all normals.\n");
	printf("       -o     only print spheres on scattered objects.\n");
	printf("       -s #   number of segments to render per sphere (default is 8) \n");
	printf("       -i     print spheres as icosahedrons (20 triangles/sphere) instead \n");
	
  exit(error);
}


//------------------------
//-- Main function
//-- Reads command line input, opens new file to write out to, prints any
//-- file errors and then calls "imod_to_obj" function on the chosen model file.
//-- If a third filename argument is specified it will call "imod_to_mtl" too.

int main( int argc, char *argv[])
{
  int i;
  FILE *fout, *fmout;
  Imod *imod;
  char *progname = imodProgName(argv[0]);
	char *matFileName;
	
  if (argc < 2) {
    imodVersion(progname);
    imodCopyright();
    usage(0);
  }
	
  for (i = 1; i < argc; i++){
    if (argv[i][0] == '-')
      switch (argv[i][1]){
      case 'l':
        lowRes = 1;
        break;
					
			case 'a':
				printAllObjs = 1;
				break;
			
			case 'r':
				rotateModel = 1;
				break;
			
			case 'i':
				useIcosahedrons = 1;
				break;
				
			case 's':
				sphereSegments = atof(argv[++i]);
				break;
				
			case 'n':
				printNormals = 1;
				break;
				
			case 'f':
				flipNormals = 1;
				break;
				
			case 'o':
				onlyScatSpheres = 1;
				break;
				
      default:
        usage(-1);
        break;

      }
    else
      break;
  }

  if (argc - i != 2 && argc - i != 3)
    usage(-1);
	
	//## OPEN IMOD FILE FOR READING:
	imod = imodRead(argv[i]);
  if (!imod){
    fprintf(stderr, "%s: Error reading imod model %s\n", argv[0], argv[i]);
    exit(3);
  }
	
	//## OPEN/CREATE .OBJ FILE FOR WRITING:
	
  fout = fopen( argv[i+1] , "w");
  if (fout == NULL){
    printf("Couldn't open output file %s.\n", argv[i+1]);
    exit(10);
  }
	
	if(argc - i == 3)
	{
		printMatFile = 1;
		matFileName = argv[i+2];
	}
	
	
  imod_to_obj(imod, fout, matFileName);
	
  fclose(fout);
	
	
	//## OPEN/CREATE .MTL FILE FOR WRITING (IF APPLICABLE):
	
	if(printMatFile)
	{
		fmout = fopen( matFileName , "w");
		if (fmout == NULL){
			printf("Couldn't open material output file %s.\n", matFileName);
			exit(10);
		}
		imod_to_mtl(imod, fmout);
	}
	
	fprintf(stdout, "Finished writing '%s' \n", argv[i] );
	fprintf(stdout, "  # objects on: %d\n", numObjs     );
	fprintf(stdout, "  # spheres:    %d\n", numSpheres  );
	fprintf(stdout, "  # vertices:   %d\n", numVertices );
	fprintf(stdout, "  # faces:      %d\n", numFaces    );
	fprintf(stdout, "  mtl file generated: %s\n", (printMatFile) ? "yes" : "no" );
	
  exit(0);
}

//------------------------
//-- Takes an imod model and outputs the meshes into a OBJ file by printing
//-- then calling "printObject" on each object.
//-- For more information about OBJ format see:
//-- > http://www.fileformat.info/format/wavefrontobj/egff.htm   (official specs)   OR
//-- > http://www.andrewnoske.com/wiki/index.php?title=OBJ_file_format  (author's notes)

void imod_to_obj(Imod *imod, FILE *fout, char *matFileName)
{
	int ob;
  fprintf(fout,"# WaveFront *.obj file (generated from an IMOD model by imod2obj)\n");
	
	if(printMatFile)
	{
		fprintf(fout,"# Material values are stored in this .mtl file:\n");
		fprintf(fout,"mtllib %s\n", matFileName);
	}
	
  fprintf(fout,"\n\n");
	for(ob = 0; ob < imod->objsize; ob++)
	{
    printObject(imod, ob, fout);
		fprintf(fout,"\n\n");
  }
	fprintf(fout,"\n\n");
	fprintf(fout,"# For more info on OBJ file format see:\n");
	fprintf(fout,"#  http://www.andrewnoske.com/wiki/index.php?title=OBJ_file_format\n");
}

//------------------------
//-- Outputs to file [fout] a unique name for the given object [ob] in the form:
//--   "obj1_Object_Name"
//-- Note that the object number is included (to ensure uniquness) and the user
//-- assigned object name is also printed, but using an underscore (_) to replace
//-- any whitespaces, brackets and several other characters which OBJ format
//-- does not allow in a single name.
//-- These dangerous characters include:    \t \n \\ # c . : () [] {} ' " + & ; |

static void printObjSafeUniqueObjectName(Imod *imod, int ob, FILE *fout )
{
	int i;
	Iobj *obj = &imod->obj[ob];
	
	fprintf( fout,"obj%d_", (ob+1) );		// write out object number as "obj1_"
	
	for(i = 0; i < IOBJ_STRSIZE; i++)		// write out object's name as "Object_Name"
	{
		char c = obj->name[i];							// get character
		if( c == 0x00 )
			break;
		else if( c==' ' || c=='\t' || c=='\n' )					// whitespace chars
			fprintf( fout,"_" );
		else if( c=='(' || c=='[' || c=='{' )												// open brackets
			fprintf( fout,"<" );
		else if( c==')' || c==']' || c=='}' )												// close brackets
			fprintf( fout,">" );
		else if( c=='\'' || c=='"' )																// quote symbols
			fprintf( fout,"`" );
		else if( c=='#' || c=='.' || c==',' || c=='\\' || c==':' )	// other
			fprintf( fout,"_" );
		else if( c=='+' || c=='&' || c==';' || c=='|' )			// symbols cinema 4D rejects
			fprintf( fout,"_" );
		else
		  fprintf( fout,"%c", c );
	}
}

//------------------------
//-- Prints out to file the OBJ text for that object by first printing
//-- the material by calling "printMaterial" function, then 
//-- calling a seperate function to "printScatContours" or "printMesh"
//-- depending on the way this appeared in code.

static void printObject(Imod *imod, int ob, FILE *fout)
{
  int co;
  Iobj *obj = &imod->obj[ob];
	int objScattered = iobjScat(obj->flags) ? 1 : 0;
  int hasSpheres   = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
	if ( !printAllObjs && iobjOff(obj->flags) )		// if object off, and "-a" not entered:
    return;																				// exit early
	
	// See if object has spheres to draw
  for (co = 0; co < obj->contsize && !hasSpheres; co++)
    if (obj->cont[co].sizes)
      hasSpheres = 1;
	
	
  //## PRINT OBJECT'S DATA:
	
	fprintf(fout,"\n");
	fprintf(fout,"g ");
	printObjSafeUniqueObjectName( imod, ob, fout );
	fprintf(fout,"\n");
	
	if(printMatFile)
	{
		fprintf(fout,"usemtl ");
		printObjSafeUniqueObjectName( imod, ob, fout );
		fprintf(fout,"\n");
	}
	fprintf(fout,"\n");
	
	if( !objScattered )
	{
    if (iobjMesh(obj->flags))
      printMesh(imod, ob, fout);
	}
  if (hasSpheres) {
		if( objScattered || !onlyScatSpheres )
			printScatContours(imod, ob, fout);
  }
	
  numObjs++;
}





//------------------------
//-- Prints out the mesh generated over the given object ("ob") in
//-- OBJ format which involves first outputting the list of all points
//-- using the "printAllPoints" function, then connecting points into triangles.

static void printMesh(Imod *imod, int ob, FILE *fout)
{
  Iobj *obj = &imod->obj[ob];
	int objFill = iobjFill(obj->flags);		// object has been set to show faces
	int objLine = iobjLine(obj->flags);		// object has been set to show lines
  Imesh *mesh;
  int m, c, p, i, j;
  float zscale = imod->zscale;
  int *ilist;
  int lsize;
  int index, ind, poly;
  Ipoint vert, norm;
  int resol;
  int listInc, vertBase, normAdd;
	int objClosed, contClosed;
	
	
  imodMeshNearestRes(obj->mesh, obj->meshsize, lowRes, &resol);
  
	
	
	//## FOR EACH MESH (TYPICALLY JUST ONE) PRINT MESH OR POINTS: 
	
  for(m = 0; m < obj->meshsize; m++)
	{
    if (imeshResol(obj->mesh[m].flag) != resol)
      continue;
		
    poly = 0;
    mesh = &obj->mesh[m];
    ilist = (int *)malloc(sizeof(int) * mesh->lsize);
    if (!ilist) {
      printf("Error allocating memory\n");
      exit(1);
    }
		
		//## OUTPUT VERTEXES (POINTS):
		int fVert = numVertices;					// first vertex in this group
		for(i = 0; i < mesh->vsize; i+=2)
		{
			vert = mesh->vert[i];
			if(rotateModel)
				fprintf(fout,"v %.5g %.5g %.5g\n",	vert.x, (vert.z * zscale), vert.y );
			else
				fprintf(fout,"v %.5g %.5g %.5g\n",	vert.x, vert.y, (vert.z * zscale) );
			numVertices++;
		}
		
		//## OUTPUT NORMALS (IF SPECIFIED):
		int fNorm = numNormals;					// first normal in this group
		if (printNormals)
		{
			fprintf(fout,"\n");
			for(i = 1; i < mesh->vsize; i+=2){
				norm = mesh->vert[i];
				imodPointNormalize(&norm);
				if(flipNormals)
					fprintf(fout,"vn %.3f %.3f %.3f\n", norm.x, norm.y, norm.z );
				else
					fprintf(fout,"vn %.3f %.3f %.3f\n", -norm.x, -norm.y, -norm.z );
				numNormals++;
			}
		}
		
		//## OUPUT FACES (TRIANGLES):
		fprintf(fout,"\n");
		for(i = 0; i < mesh->lsize; i++)
		{
			if (mesh->list[i] == IMOD_MESH_END)
				break;
			if ( !imodMeshPolyNormFactors(mesh->list[i], &listInc, &vertBase, &normAdd) )
				continue;
			
			// Copy the indices to the new index array numbered from 0
			lsize = 0;
			i++;
			while (mesh->list[i] != IMOD_MESH_ENDPOLY)
			{
				ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
				ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
				ilist[lsize++] = mesh->list[i+vertBase]/2;			i += listInc;
			}
			
			if (!lsize)
				continue;
			
			//## OUTPUT FACESET:   (if showing surface, put out face set)
			
			for(index = 0; index < lsize / 3; index++)
			{
				ind = 3 * index;
				if(printNormals)
				{
					fprintf(fout,"f %d//%d %d//%d %d//%d\n", 
									ilist[ind+2]+fVert+1, ilist[ind+2]+fNorm+1, 
									ilist[ind+1]+fVert+1, ilist[ind+1]+fNorm+1,
									ilist[ind]  +fVert+1, ilist[ind]  +fNorm+1 );
				}
				else
				{
					fprintf(fout,"f %d %d %d\n", 
											  ilist[ind+2]+fVert+1, ilist[ind+1]+fVert+1, ilist[ind]+fVert+1);
				}
				numFaces++;
			}
		}
		
		//## FREE ANY MEMORY USED:
    if (ilist) 
      free(ilist);
  }
}


//------------------------
//-- Prints a scattered point object (in OBJ format) by printing a sphere for 
//-- every point in the object which has a radius.

static void printScatContours(Imod *imod, int ob, FILE *fout)
{
	Iobj *obj = &imod->obj[ob];
  Icont *cont;
  int hasSpheres = (iobjScat(obj->flags) || obj->pdrawsize) ? 1 : 0;
  int c, p;
  float zscale = imod->zscale;
  float radius;
  Ipoint pt;
	
	fprintf(fout,"# obj%d SPHERES:\n\n", ob+1);
	
  //## FOR EACH CONTOUR: PRINT EVERY POINT WHICH HAS A SPHERE SIZE:
	
	for(c = 0; c < obj->contsize; c++){
    cont = &(obj->cont[c]);
				
    if (!cont || !cont->psize								// if contour is empty
				|| (!hasSpheres && !cont->sizes))		// or has no sphere sizes:
      continue;																// skip this contour
		
		for (p = 0; p < cont->psize; p++)
		{
			fprintf(fout,"#   cont %d pt %d\n", c+1, p+1 );
			fprintf(fout,"g obj_%d_cont_%d_pt_%d\n", ob+1, c+1, p+1 );
			
			pt.x = cont->pts[p].x;
			pt.y = cont->pts[p].y;
			pt.z = cont->pts[p].z * zscale;
			
			if(rotateModel)
			{
				pt.y = cont->pts[p].z * zscale;
				pt.z = cont->pts[p].y;
			}
			
      radius = imodPointGetSize(obj, cont, p);
			
			if( useIcosahedrons )
				printSphereAsIcosahedron(pt, radius, fout);
			else
				printSphere( pt, radius, sphereSegments, fout );
			
			numSpheres++;
    }
  }
	fprintf(fout,"\n");	
}


//------------------------
//-- Prints a sphere as a "standard sphere" triangular mesh with the specified
//-- number of segments. This standard sphere works a lot like longditude and
//-- latitude whereby the top and bottom point connect with triangles to
//-- their adjacent points, while all other points connect together with
//-- rectangles.

static void printSphere(Ipoint pt, float radius, int segments, FILE *fout)
{
	int p, s, i, j;
	
	if(segments < 4)
		return;
	
	float DEGS_TO_RAD = 3.14159f/180.0f;
	
	int nPitch = segments / 2 + 1;
	float pitchInc = (180. / (float)nPitch  ) * DEGS_TO_RAD;
	float segInc   = (360. / (float)segments) * DEGS_TO_RAD;
	
	//## PRINT VERTEXES:
	
	fprintf(fout,"v %.5g %.5g %.5g\n", pt.x, pt.y+radius, pt.z );		// top vertex
	fprintf(fout,"v %.5g %.5g %.5g\n", pt.x, pt.y-radius, pt.z );		// bottom vertex
	numVertices += 2;
	
	int fVert = numVertices;		// index of first intermediate vertex
	for(p=1; p<nPitch; p++)			// generate all "intermediate vertices":
	{
		float out = fabs( radius * sin( (float)p * pitchInc ) );
		float y   = radius * cos( (float)p * pitchInc );
		
		for(s=0; s<segments; s++)
		{
			float x = out * cos( (float)s * segInc );
			float z = out * sin( (float)s * segInc );
			
			fprintf(fout,"v %.5g %.5g %.5g\n", x+pt.x, y+pt.y, z+pt.z );
			numVertices++;
		}
	}
	fprintf(fout,"\n");
	
	
	//## PRINT NORMALS:
	
	long fNorm;				// index of first intermediate normal
	if(printNormals)
	{
		fprintf(fout,"vn 0.0 1.0 0.0\n"  );			// top normal
		fprintf(fout,"vn 0.0 -1.0 0.0\n" );		  // bottom normal
		numNormals += 2;
		
		fNorm = numNormals;
		for(p=1; p<nPitch; p++)			// generate all "intermediate normals":
		{
			float outN = fabs(1.0f * sin( (float)p * pitchInc ));
			float yN   = cos( (float)p * pitchInc );
			
			for(s=0; s<segments; s++)
			{
				float xN = outN * cos( (float)s * segInc );
				float zN = outN * sin( (float)s * segInc );
				
				if(zN > -0.001f && zN < 0.001f)	zN = 0;
				if(xN > -0.001f && xN < 0.001f)	xN = 0;
				
				fprintf(fout,"vn %.5g %.5g %.5g\n", xN, yN, zN );
				numNormals++;
			}
		}
		fprintf(fout,"\n");
	}
	
	
	
	//## PRINT SQUARE FACES BETWEEN INTERMEDIATE POINTS:
	
	for(p=1; p<nPitch-1; p++)
	{
		for(s=0; s<segments; s++)
		{
			i = p*segments + s;
			j = (s==segments-1) ? i-segments : i;
			
			if(printNormals)
			{
				fprintf(fout,"f %d//%d %d//%d %d//%d %d//%d\n", 
								(i+1-segments)+fVert, (i+1-segments)+fNorm, 
								(j+2-segments)+fVert, (j+2-segments)+fNorm, 
								(j+2)+fVert,          (j+2)+fNorm, 
								(i+1)+fVert,          (i+1)+fNorm );
			}
			else
			{
				fprintf(fout,"f %d %d %d %d\n", 
								(i+1-segments)+fVert, (j+2-segments)+fVert, (j+2)+fVert, (i+1)+fVert );
			}
			numFaces++;
		}
	}
	
	//## PRINT TRIANGLE FACES CONNECTING TO TOP AND BOTTOM VERTEX:
	
	int offLastVerts  = fVert + (segments * (nPitch-2) );
	int offLastNorms  = fNorm + (segments * (nPitch-2) );
	for(s=0; s<segments; s++)
	{
		j = (s==segments-1) ? -1 : s;
		if(printNormals)
		{
			fprintf(fout,"f %d//%d %d//%d %d//%d\n",
							fVert-1,      fNorm-1,
							(j+2)+fVert,  (j+2)+fNorm,
							(s+1)+fVert,  (s+1)+fNorm );
			fprintf(fout,"f %d//%d %d//%d %d//%d\n",
							fVert,               fNorm,
							(s+1)+offLastVerts,  (s+1)+offLastNorms,
							(j+2)+offLastVerts,  (j+2)+offLastNorms );
		}
		else
		{
			fprintf(fout,"f %d %d %d\n", fVert-1, (j+2)+fVert,        (s+1)+fVert        );
			fprintf(fout,"f %d %d %d\n", fVert,   (s+1)+offLastVerts, (j+2)+offLastVerts );
		}
		numFaces = numFaces + 2;
	}
	
	fprintf(fout,"\n");	
}


//------------------------
//-- Prints a sphere as an isohedron - a regular polyedron with 20 identical
//-- equilateral triangular faces, 30 edges and 12 vertices.
//-- Code was modified from:
//-- http://www.csee.umbc.edu/~squire/reference/polyhedra.shtml#icosahedron

static void printSphereAsIcosahedron(Ipoint pt, float radius, FILE *fout)
{
	int i;
	Ipoint v[12];
	
	v[0].x = 0.000;			v[0].y = 1.000;			v[0].z = 0.000;			// top-most point
	v[1].x = 0.894;			v[1].y =  0.447; 		v[1].z = 0.000;
	v[2].x = 0.276;			v[2].y =  0.447;		v[2].z = 0.851;
	v[3].x = -0.724;		v[3].y =  0.447;		v[3].z = 0.526;
	v[4].x = -0.724;		v[4].y =  0.447;		v[4].z = -0.526;
	v[5].x = 0.276;			v[5].y =  0.447;		v[5].z = -0.851;
	v[6].x = 0.724;			v[6].y = -0.447;		v[6].z = 0.526;
	v[7].x = -0.276;		v[7].y = -0.447;		v[7].z = 0.851;
	v[8].x = -0.894;		v[8].y = -0.447;		v[8].z = 0.000;
	v[9].x = -0.276;		v[9].y = -0.447;		v[9].z = -0.851;
  v[10].x= 0.724;			v[10].y= -0.447;		v[10].z= -0.526;
  v[11].x= 0.000;			v[11].y= -1.000;		v[11].z= 0.000;			// bottom-most point
	
	//## PRINT VERTEXES:
	
	for(i=0; i<12; i++)
	{
		fprintf(fout,"v %.5g %.5g %.5g\n",
						v[i].x*radius+pt.x,
						v[i].y*radius+pt.y,
						v[i].z*radius+pt.z );
		numVertices++;
	}
	fprintf(fout,"\n");	
	
	//## PRINT FACES:
	
	fprintf(fout,"f -12 -10 -11\n");		// |-- top-most triangles
  fprintf(fout,"f -12 -9 -10\n");			// |
  fprintf(fout,"f -12 -8 -9\n");			// |
  fprintf(fout,"f -12 -7 -8\n");			// | 
  fprintf(fout,"f -12 -11 -7\n");			// |
  fprintf(fout,"f -1 -6 -5\n");			// |-- bottom-most triangles
  fprintf(fout,"f -1 -5 -4\n");			// |
  fprintf(fout,"f -1 -4 -3\n");			// |
  fprintf(fout,"f -1 -3 -2\n");			// |
  fprintf(fout,"f -1 -2 -6\n");			// |
  fprintf(fout,"f -11 -10 -6\n");			// |-- downwards pointing
  fprintf(fout,"f -10 -9 -5\n");			// |   triangles
  fprintf(fout,"f -9 -8 -4\n");				// |
  fprintf(fout,"f -8 -7 -3\n");				// |
  fprintf(fout,"f -7 -11 -2\n");			// |
  fprintf(fout,"f -6 -10 -5\n");		// |-- upwards pointing
  fprintf(fout,"f -5 -9 -4\n");			// |   triangles
  fprintf(fout,"f -4 -8 -3\n");			// |
  fprintf(fout,"f -3 -7 -2\n");			// |
  fprintf(fout,"f -2 -11 -6\n");		// |
	numFaces = numFaces + 20;
	
	fprintf(fout,"\n");
}



//------------------------
//-- Takes an imod model and outputs the meshes into a Wavefront's MTL
//-- (Material Template Library) file by calling "printMaterial" on each object.
//-- For more information about MTL format see:
//-- > http://en.wikipedia.org/wiki/Material_Template_Library   (official specs)

void imod_to_mtl(Imod *imod, FILE *fout)
{
	int ob;
  fprintf(fout,"# WaveFront *.mtl file (generated from an IMOD model by imod2obj)\n");
	
  fprintf(fout,"\n\n");
	for(ob = 0; ob < imod->objsize; ob++)
    printMaterial(imod, ob, fout);
	fprintf(fout,"\n\n");
	fprintf(fout,"# For more info on MTL file format see:\n");
	fprintf(fout,"#  http://en.wikipedia.org/wiki/Material_Template_Library");
}

//------------------------
//-- Prints out the Material Template Library code representing the 
//-- material properties of the specified object ("ob").

static void printMaterial(Imod *imod, int ob, FILE *fout)
{
	Iobj *obj = &imod->obj[ob];
  float ambient   = ((int)obj->ambient)   / 255.;
	float diffuse   = ((int)obj->diffuse)   / 255.;
  float specular  = ((int)obj->specular)  / 255.;
  float shininess = ((int)obj->shininess) / 255.;
  float r = (float)obj->red;
  float g = (float)obj->green;
  float b = (float)obj->blue;
	
  if (shininess < 0.01)
    shininess= 0.01;
	
  //## PRINT MATERIAL 
	//## CAN REFERENCE THIS TAG:
	
	fprintf(fout,"\n");
	fprintf(fout,"#MATERIAL FOR OBJECT %d:\n", (ob+1) );
  fprintf(fout,"newmtl ");
	printObjSafeUniqueObjectName(imod, ob, fout );
	fprintf(fout,"\n");
	fprintf(fout,"Ka %g %g %g\n", r*ambient, g*ambient, b*ambient);
  fprintf(fout,"Kd %g %g %g\n", r*diffuse, g*diffuse, b*diffuse);
  fprintf(fout,"Ks %g %g %g\n", r*specular, g*specular, b*specular);
  fprintf(fout,"Ns %g\n", shininess*1000.);
  fprintf(fout,"d %g\n",  obj->trans/100.0f);		//|-- transparency
	fprintf(fout,"Tr %g\n", obj->trans/100.0f);		//|
  fprintf(fout,"\n");
}



/*
 $Log: imod2obj.cpp,v $
 Revision 1.0  2011/12/22 21:37:14  noske
 Modified file from "imod2vrml.c" to output OBJ format
 */
